### ISSUE TITLE
KeyVault Schema Requires Access Policy - Docs Say It Is Not Required

### ISSUE THREAD

**Issue by austincrft** [Azure/azure-resource-manager-schemas#521]
In the [Microsoft.KeyVault/vaults docs](https://docs.microsoft.com/en-us/azure/templates/microsoft.keyvault/vaults#AccessPolicyEntry), it says that `accessPolicies` is not required:

![image](https://user-images.githubusercontent.com/3616544/47117070-fbcf6d00-d228-11e8-958b-faa45d8af42b.png)

When I deploy with this resource, I get no errors

```
{
    "type": "Microsoft.KeyVault/vaults",
    "name": "[variables('keyVaultName')]",
    "apiVersion": "2018-02-14",
    "location": "[resourceGroup().location]",
    "tags": {},
    "scale": null,
    "properties": {
        "sku": "[parameters('keyVaultSku')]",
        "tenantId": "[parameters('tenantId')]",
        "accessPolicies": [],
        "enabledForDeployment": false,
        "enabledForDiskEncryption": false,
        "enabledForTemplateDeployment": false
    },
    "dependsOn": []
}
```

When I deploy with this resource (`accessPolicies` missing)

```
{
    "type": "Microsoft.KeyVault/vaults",
    "name": "[variables('keyVaultName')]",
    "apiVersion": "2018-02-14",
    "location": "[resourceGroup().location]",
    "tags": {},
    "scale": null,
    "properties": {
        "sku": "[parameters('keyVaultSku')]",
        "tenantId": "[parameters('tenantId')]",
        "enabledForDeployment": false,
        "enabledForDiskEncryption": false,
        "enabledForTemplateDeployment": false
    },
    "dependsOn": []
}
```

I get this error:

![image](https://user-images.githubusercontent.com/3616544/47117403-dabb4c00-d229-11e8-83d1-b90a8709a883.png)

**Comment by WillParry** [comment]
I've seen this same issue and raised via twitter and msdn forums [https://social.msdn.microsoft.com/Forums/en-US/dd14a115-922b-415c-9c81-bd423c13020f/how-do-i-deploy-keyvault-from-arm-template-without-overwriting-accesspolicies?forum=AzureKeyVault](https://social.msdn.microsoft.com/Forums/en-US/dd14a115-922b-415c-9c81-bd423c13020f/how-do-i-deploy-keyvault-from-arm-template-without-overwriting-accesspolicies?forum=AzureKeyVault )

**Comment by michaelwstark** [comment]
Just ran into this as well.  It is blocking our ability to leverage managed service identity as the dynamically added identities are reset with "accessPolicies" set, and it errors if the value is omitted.  There does not appear to be a way to incrementally deploy a key vault without impacting the access policies if they are not statically defined.

**Comment by austincrft** [comment]
@michaelwstark, a workaround is to use the Azure PowerShell library to get the current state of access policies and pass them in as a parameter to the deployment template.

An example of this can be found here: https://toonvanhoutte.wordpress.com/2018/11/13/key-vault-deployment-removes-access-policies/

This is doing slightly more by setting Azure DevOps variables, but you should be able to modify it for your needs.

**Comment by Soruk** [comment]
Any news on this issue?
The documentation is still mentioning that the "accessPolicies" are not required but the Azure Resource Template deployment gives still the same error:
"BadRequest" "message": "An invalid value was provided for 'accessPolicies'." 
With this issue, I have to execute PowerShell script to recreate aditionnal access policies after each deployment 😩

**Comment by seanbamsft** [comment]
The template schema and documentation are automatically generated from the REST API spec. In the REST API this property is optional because it can be omitted when recovering a soft-deleted vault (the existing access policy will be restored). The docs have been updated to indicate that this field is only optional when `createMode` is `recover`. The access policies are also optional when you use `PATCH` instead of `PUT`. We discussed the possibility of adding patch semantics to the `PUT` API for template deployments, but the proposal was rejected by the Azure Resource Manager team. If you know the vault already exists you can use the `reference()` function to get the current access policies and append an entry. The ARM team had a proposed feature that would let you handle the case where the vault didn't exist, but I don't think it's possible with the current template language.

**Comment by mslot** [comment]
So on the one side we have the "Use ARM to automatically deploy all things", and on the other side we have "be secure, use keyvault, no extra code needed", but we cant use both.

I need to have managed identites added dynamically. I want to be able to redeploy my keyvault and not having access policies wiped.

I am leaving a comment here, so I get updates! Please tell me when this is fixed, or if it will be fixed. I see this as a bug, and not a feature.

**Comment by wouterroos** [comment]
@mslot Agreed, it's unacceptable that such a seemingly common use-case can't be achieved using ARM.

**Comment by mslot** [comment]
> @mslot Agreed, it's unacceptable that such a seemingly common use-case can't be achieved using ARM.
> For anyone having the same issue, the closest I came to an acceptable solution is making the value of both the `accessPolicies` and `create` mode conditional based on a parameter that specifies if the key vault already exists withing the target resource group (see below). This allows you to either set this value manually before deploying the template for the first time to a given resource group or determine if the resource already exists using some external script. When the parameter is set to `true', the template will use the `recover`mode which makes the`accessPolicies`parameter optional, allowing you to pass`null` which will not remove any existing access policies when deploying the template.
> 
> ```
> "accessPolicies": "[if(parameters('My.KeyVault.existsWithinTargetResourceGroup'), json('null'), json('[]'))]",
> "createMode": "[if(parameters('My.KeyVault.existsWithinTargetResourceGroup'), 'recover', 'create')]",
> ```

This I have thought about. My solution is to completely remove the keyvault from the ARM template, and create it though a script ONLY if it doesn't exists. I run this script before I run my ARM that sets up my infrastructure.

**Comment by seanbamsft** [comment]
@wouterroos That's something of an abuse of the createMode parameter. It happens to work because we return a 200 if you attempt to recover a vault that hasn't been deleted yet and we aren't enforcing the absence of other properties during recovery. But if you are going to have a parameter that indicates whether the vault exists or not, you could simply make the [entire vault conditional](https://docs.microsoft.com/en-us/azure/architecture/building-blocks/extending-templates/conditional-deploy) on that parameter without relying on this behavior.

In general, the direction ARM is going in is to support [DataActions in ARM RBAC policies](https://docs.microsoft.com/en-us/azure/role-based-access-control/role-definitions) rather than have each service provide its own access control mechanism. If Key Vault were to support data actions the same way Storage does, it would remove the need to update the vault to grant access to its contents.

**Comment by winperec** [comment]
Very weird behavior, it makes ARM templates for KeyVault almost useless without external scripts...

**Comment by onionhammer** [comment]
Weird that this was just closed without any comment at all..

**Comment by ahughes257** [comment]
Still an issue getting the same badrequest when not specifying this optional parameter... 

2021-11-01T12:09:37.8684681Z ##[error]BadRequest: The parameter accessPolicies is not specified.

**Comment by anthony-c-martin** [comment]
I think this issue was most likely closed as it's not technically a schemas issue - as @seanbamsft mentioned, there are legitimate scenarios where `accessPolicies` can be `null`, which is why it's non-required in the schema.

@bmoore-msft / @alex-frankel - do we have any guidance to share on the 'best practice' for managing a KeyVault with IaC, and avoiding the issue with `accessPolicies` being wiped out?

Reopening this issue for now to track the discussion.

**Comment by bmoore-msft** [comment]
For IaC I'm not sure we have a good option/pattern, if accessPolicies is required for a declarative PUT - aside from declaring the actual accessPolicies you want.  If accessPolicies is optional on ```recover```, I wonder if it can be made optional on ```createMode=default```.  Not sure what the implications of that would be ( @seanbamsft  ) but that seems to be what the thread is suggesting...

**Comment by zapadoody** [comment]
Simply including "accessPolicies": [] is a perfect workaround for me. It seems like a documentation issue, maybe update the documentation `accessPolicies` as required but nullable.

**Comment by feliasson** [comment]
> Simply including "accessPolicies": [] is a perfect workaround for me. It seems like a documentation issue, maybe update the documentation `accessPolicies` as required but nullable.

No, setting that empty array would wipe any access policies already there until applied again, that's the whole issue basicly, we are required to input the accessPolicies and therefor wiping anything there. That in combination with no real support for checking if resources are already existent within ARM/bicep makes this a huge issue. 
 
ANY UPDATES SOON?

**Comment by bmoore-msft** [comment]
KeyVault now supports data plane RBAC: 

https://docs.microsoft.com/en-us/azure/key-vault/general/rbac-guide?tabs=azure-cli 

So this not only creates another scenario where accessPolicies:[] is not required, but separates vaults and permissions, so you don't have the consequences of specifying an empty array on a subsequent deployment.

**Comment by michaelwstark** [comment]
My team moved to RBAC keyvaults specifically to avoid this problem while building out our new services.  Doing so completely mitigated the problem I described earlier in the thread.
